File: src/movement.js
/**
 * 键盘移动方向合成：把当前按住的方向动作合成一步位移。
 * 支持斜向（如 左+上 → 左上）、相反方向抵消（上+下 → 不动）、边界裁剪。
 * 同时被渲染进程（<script> 全局 combineDirections）和测试（require）使用。
 */
;(function (root, factory) {
  if (typeof module === 'object' && module.exports) {
    module.exports = factory()
  } else {
    root.combineDirections = factory()
  }
})(typeof self !== 'undefined' ? self : this, function () {
  const DIRS = {
    moveUp: { dr: -1, dc: 0 },
    moveDown: { dr: 1, dc: 0 },
    moveLeft: { dr: 0, dc: -1 },
    moveRight: { dr: 0, dc: 1 }
  }

  /**
   * @param {Set<string>} heldActions 当前按住的方向动作（如 moveUp / moveLeft）
   * @param {number} rows 棋盘行数
   * @param {number} cols 棋盘列数
   * @param {number} row 光标当前行
   * @param {number} col 光标当前列
   * @returns {{row: number, col: number}|null} 移动后的位置；无有效方向时返回 null
   */
  function combineDirections(heldActions, rows, cols, row, col) {
    let dr = 0
    let dc = 0
    for (const action of heldActions) {
      const d = DIRS[action]
      if (d) {
        dr += d.dr
        dc += d.dc
      }
    }
    if (dr === 0 && dc === 0) return null
    return {
      row: Math.max(0, Math.min(rows - 1, row + dr)),
      col: Math.max(0, Math.min(cols - 1, col + dc))
    }
  }

  return combineDirections
})

File: android/core/src/main/kotlin/com/wpiaopiao/saolei/core/Movement.kt
package com.wpiaopiao.saolei.core

/**
 * 方向合成：把当前按住的方向动作合成一步位移。
 * 支持斜向（左+上 → 左上）、相反方向抵消（上+下 → 不动）、边界裁剪。
 * 由桌面版 src/movement.js 移植。
 */
object Movement {

    private val DIRS: Map<String, Pair<Int, Int>> = mapOf(
        "moveUp" to (-1 to 0),
        "moveDown" to (1 to 0),
        "moveLeft" to (0 to -1),
        "moveRight" to (0 to 1)
    )

    /**
     * @param held 当前按住的方向动作（moveUp/moveLeft 等）
     * @return 移动后的 (row, col)；无有效方向时返回 null
     */
    fun combineDirections(
        held: Set<String>,
        rows: Int,
        cols: Int,
        row: Int,
        col: Int
    ): Pair<Int, Int>? {
        var dr = 0
        var dc = 0
        for (action in held) {
            val d = DIRS[action]
            if (d != null) {
                dr += d.first
                dc += d.second
            }
        }
        if (dr == 0 && dc == 0) return null
        return (row + dr).coerceIn(0, rows - 1) to (col + dc).coerceIn(0, cols - 1)
    }
}
